### PR TITLE
dnsdist: Add `getAddressInfo()` for asynchronous DNS resolution

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

Closes https://github.com/PowerDNS/pdns/issues/9875

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/dnsdistdist/docs/reference/config.rst [hunk 1]
```diff
@@ -1812,6 +1812,25 @@ These values can be set at configuration time via:
 Other functions
 ---------------
 
+.. function:: getAddressInfo(hostname, callback)
+
+  .. versionadded:: 1.9.0
+
+  Asynchronously resolve, via the system resolver (using ``getaddrinfo()``), the supplied ``hostname`` to IPv4 and IPv6 before invoking the supplied ``callback`` function with the ``hostname`` and a list of IPv4 and IPv6 addresses as :class:`ComboAddress`.
+  For example, to get the addresses of Quad9's and dynamically add them as backends:
```
Comment (Habbie) [inline pdns/dnsdistdist/docs/reference/config.rst:1820]:
```suggestion
  For example, to get the addresses of Quad9's resolver and dynamically add them as backends:
```

Comment (rgacogne) [reply pdns/dnsdistdist/docs/reference/config.rst:1820]: Oops, thanks!

**Comment by Habbie** [comment]
After we merge this, we should update `docker/dnsdist*` to use it, and drop the python file.

File: pdns/dnsdistdist/dnsdist-resolver.cc [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+#include "iputils.hh"
+
+namespace dnsdist::resolver
+{
+void asynchronousResolver(const std::string& hostname, const std::function<void(const std::string& hostname, std::vector<ComboAddress>& ips)>& callback)
+{
+  addrinfo hints{};
+  hints.ai_family = AF_UNSPEC;
+  hints.ai_flags = AI_V4MAPPED | AI_ADDRCONFIG;
+  hints.ai_socktype = SOCK_DGRAM;
```
Comment (omoerbeek) [inline pdns/dnsdistdist/dnsdist-resolver.cc:34]: you should also set `ai_protocol`
`AI_V4MAPPED` might not be generally available

Comment (rgacogne) [reply pdns/dnsdistdist/dnsdist-resolver.cc:34]: Good point about `AI_V4MAPPED`, in fact I don't think we actually want to set it (it's the default on Linux but not everywhere indeed).
I'm genuinely curious as to what is the point of setting `ai_protocol`, though?

**Comment by rgacogne** [comment]
Rebased on master to fix a conflict.

File: pdns/dnsdistdist/docs/reference/config.rst [hunk 3]
```diff
@@ -1915,6 +1915,25 @@ These values can be set at configuration time via:
 Other functions
 ---------------
 
+.. function:: getAddressInfo(hostname, callback)
+
+  .. versionadded:: 1.9.0
+
+  Asynchronously resolve, via the system resolver (using ``getaddrinfo()``), the supplied ``hostname`` to IPv4 and IPv6 before invoking the supplied ``callback`` function with the ``hostname`` and a list of IPv4 and IPv6 addresses as :class:`ComboAddress`.
```
Comment (omoerbeek) [inline pdns/dnsdistdist/docs/reference/config.rst:1922]: We use `AI_ADDRCONFIG`, so we only get A if v4 is configured and AAAA if v6. 
```suggestion
  Asynchronously resolve, via the system resolver (using ``getaddrinfo()``), the supplied ``hostname`` to IPv4 and IPv6 addresses (if configured on the host) before invoking the supplied ``callback`` function with the ``hostname`` and a list of IPv4 and IPv6 addresses as :class:`ComboAddress`.
```

**Review by omoerbeek — APPROVED** [review]
One small nit

**Comment by bondar-aleksandr** [comment]
team, how can I use this feature to dynamically refresh list of real-servers?

**Pull request merged**